### PR TITLE
Feature #175

### PR DESCRIPTION
**Pull request by aleneum**
* `Machine.before_state_change` and `Machine.after_state_change` will not be appended with `Transition.before/after` any longer.
   - this allows to change `before_state_change` and `after_state_change` during runtime (as required for #170).
* The order of `Machine.before_state_change` and `Transition.before` has been reversed. This way `Machine` methods surround `Transition` methods.
* New constructor parameter `prepare_conditions_check` (temporary name) has been introduced
    - represents a `prepare` for transitions which is called *once* for every triggered event independent of how many transitions might be possible, checked and if these transitions are successful.

### CODE REVIEW

**Comment by aleneum** [comment]
Without objection/feedback, I will merge this in a week. This will also be the moment to update the changelog and tag 0.4.4 for a release.

**Comment by ankostis** [comment]
Not an objection, but a wish: 
any thoughts about adding a `finalize_after_all_transitions` in a finally block that fires once, both for successful/failed transitions?

Should I open a new issue to discuss ramifications, such as, where to store collected errors. etc?

**Comment by aleneum** [comment]
> any thoughts about adding a finalize_after_all_transitions in a finally block that fires once, both for successful/failed transitions?

right, I will add this

> Should I open a new issue to discuss ramifications, such as, where to store collected errors. etc?

Please do so. Catching and managing errors is (currently) out of `transitions` scope.

**Comment by ankostis** [comment]
Created #186 for `finalize_after_all_transitions`.

**Comment by ankostis** [comment]
Since you named it `finalize_event`, why not `prepare_event`, to promote the relationship?

**Comment by aleneum** [comment]
> Since you named it finalize_event, why not prepare_event, to promote the relationship?

I thought about this (or something like `on\before_event`) but this might make some people believe it is _always_ executed. But it is only executed if the transition is technically possible (if `state.name in self.transition`). Maybe I am overthinking this though. Phew, naming is tough 😓.

**Comment by ankostis** [comment]
So the prepare/finalize global methods are never executed for "final" states, i.e. those without any outgoing transitions on them.
Ok, that's the expected behavior, I guess.
Or am I missing something here?

Is it possible to have an event, even for a "final" state?

PS: Naming is the hardest part in programming.
But it is fun :-)

**Comment by aleneum** [comment]
> So the prepare/finalize global methods are never executed for "final" states, i.e. those without any outgoing transitions on them.

nope, not right now.

> Is it possible to have an event, even for a "final" state?

well, since we are talking about a feature in development, it's up to us how prepare/finalize behave. We could just extend the try/except block and trigger the prepare and finalize callbacks all the time. In this case I would probably name them `before/after_event` since I feel this is the most consistent scheme considering `before/after_state_change`.

What do you think?

**Comment by ankostis** [comment]
I would expect that for a "final" state, no code should run at all, because there is nothing the user can do to change that.  I mean, it is not that he can try many triggers until the correct one fires - no, there is no correct, trigger, that is a dead-end.  So running code in `prepare/finalize` would be an undesirable side affect, like `opening a db-connection to write empty.

If you care to add such functionality, it might be better to have separate global callback, like `on_dead_trigger` or something like that.   
But that's my 2cnt on this.



Regarding the beauty contest, `before/after` is smooth, but `prepare/finalize`  has the connotation that the last one gets executed regardless of failures.  
And I like the `_event` suffix because it is not related to a single transition, but all those involved (but I guess that's slightly different meaning of the term "event" than yours).

(I hope I'm not intervening too much to your mental space)

**Comment by aleneum** [comment]
> I mean, it is not that he can try many triggers until the correct one fires - no, there is no correct, trigger, that is a dead-end. 

Good point. So I guess we should leave it like it is and only process callbacks if a transition is actually possible.

> prepare/finalize has the connotation that the last one gets executed regardless of failures.

agreed. Also `prepare_transition` does not imply that a transition will actually happen (like `before_transition`) 

> And I like the _event suffix because it is not related to a single transition, but all those involved.

I refer to the underlying `Event` object. Thats why `prepare_event` would not be accurate. But `somewhere_in_event` is no catchy at all. `prepare_transition` because we have already determined that a transition is possible but we do not know which one yet.

**Comment by ankostis** [comment]
By the way, I hadn't  thought of the dead-trigger subtlety.
Now I'm thinking that the proper way to deal with it is to invoke a *single* method on machine(or model), called like `triggered_on_final_state()` that, by default, throws `MachineError`.

Actually I was bitten by this problem while developing my model.  
You see, I'm using YAML-text to setup my transitions, and I had some comma in the wrong place, so an outgoing transitions wasn't declared to the state I wanted (that state was in effect "final"), and I couldn't understand why I was getting `false`, although all conditions were supposed to be `true`.
A `MachineError` would have waked me up.

**Comment by ankostis** [comment]
Setting `ignore_invalid_triggers=False` does the job also for final-states. 
Please ignore my last comment!  It It was late at night and never examine carefully what was my issue back then.

**Comment by aleneum** [comment]
if nobody objects, I will merge this next week. Also this will be the moment to update the manual and the changelog and also tag version 0.4.4.

**Comment by wtgee** [comment]
:+1:

File: transitions/core.py [hunk 1]
```diff
@@ -284,13 +285,27 @@ def _trigger(self, model, *args, **kwargs):
...
+        for func in self.machine.prepare_event:
+            self.machine._callback(func, event_data)
+            logger.debug("Executed machine preparation callback '%s' before conditions." % func)
+
+        result = False
+        try:
+            for t in self.transitions[state.name]:
+                event_data.transition = t
+                if t.execute(event_data):
+                    result = True
```
Comment (ankostis) [inline transitions/core.py]: Can we store the `result` on the `event_data.result` so `finalize_event` knows that?
Do we have time before 4.4 release?

**Pull request merged**